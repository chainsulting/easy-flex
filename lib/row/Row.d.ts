import React from 'react';
import { BaseFlexProps } from '../baseFlex/BaseFlex';
import { Falsifiable, FlipDirection, ViewportThreshold } from '../types';
export interface RowProps extends Omit<BaseFlexProps, 'direction'> {
    /** Flips the content in the direction set by flipDirection. */
    flip?: boolean;
    /** Sets what happens if the content shall be flipped. */
    flipDirection?: Falsifiable<FlipDirection>;
    /** Sets the viewport threshold. The content will be flipped if the viewport's width is smaller than the threshold. If no threshold is set, the default threshold is used. */
    viewportThreshold?: Falsifiable<ViewportThreshold>;
}
export declare const Row: React.ForwardRefExoticComponent<RowProps & React.RefAttributes<HTMLDivElement>>;
