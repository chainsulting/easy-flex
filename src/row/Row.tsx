import React, { forwardRef, useMemo } from 'react';
import { BaseFlex, BaseFlexProps } from '../baseFlex/BaseFlex';
import { FlexDirection, FlipDirection, ViewportThreshold } from '../types';
import { getViewportThreshold, useDimension, useEasyFlexTheme } from '../utils/base';

export interface RowProps extends Omit<BaseFlexProps, 'flexDirection'> {
	/** Flips the content in the direction set by flipDirection. */
	flip?: boolean;
	/** Sets what happens if the content shall be flipped. */
	flipDirection?: FlipDirection;
	/** Sets the viewport threshold. The content will be flipped if the viewport's width is smaller than the threshold. If no threshold is set, the default threshold is used. */
	viewportThreshold?: ViewportThreshold | number;
}

export const Row = forwardRef<HTMLDivElement, RowProps>(
	({ children, flip, flipDirection, viewportThreshold, ...props }, ref) => {
		const theme = useEasyFlexTheme();
		const { width } = useDimension();

		const flexDirection = useMemo<FlexDirection>(() => {
			if (
				flipDirection !== undefined &&
				(flip ||
					(flip === undefined &&
						(viewportThreshold !== undefined
							? width < getViewportThreshold(theme, viewportThreshold)
							: width < theme.viewport.defaultThreshold)))
			) {
				switch (flipDirection) {
					case 'flip':
						return 'column';
					case 'reverse':
						return 'row-reverse';
					case 'flip-reverse':
						return 'column-reverse';
				}
			}
			return 'row';
		}, [flip, flipDirection, theme, viewportThreshold, width]);

		return (
			<BaseFlex flexDirection={flexDirection} ref={ref} {...props}>
				{children}
			</BaseFlex>
		);
	}
);

Row.displayName = 'Row';
