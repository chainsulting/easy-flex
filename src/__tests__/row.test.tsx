import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { Row } from '../row/Row';

describe('<Row />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<Row />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
