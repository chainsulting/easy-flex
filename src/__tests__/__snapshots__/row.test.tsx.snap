// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> should have the basic css settings 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  padding-top: 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<body>
  <div>
    <div
      class="c0"
      data-direction="row"
      data-margin-bottom="0px"
      data-margin-left="0px"
      data-margin-right="0px"
      data-margin-top="0px"
      data-padding-bottom="0px"
      data-padding-left="0px"
      data-padding-right="0px"
      data-padding-top="0px"
    />
  </div>
</body>
`;
